### PR TITLE
HTS-1145: Added functionality for privileged access

### PR DESCRIPTION
**Pull request by Jackie1988**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavetestadminfrontend/controllers/HelpToSaveApiController.scala [hunk 1]
```diff
@@ -164,7 +173,8 @@ class HelpToSaveApiController @Inject()(http: WSHttp, authConnector: AuthConnect
                    |    "surname" : "${params.requestBody.surname}",
                    |    "dateOfBirth" : "${params.requestBody.dateOfBirth}",
                    |    "contactDetails" : $getContactDetailsJson,
-                   |    "registrationChannel" : "${params.requestBody.registrationChannel}"
+                   |    "registrationChannel" : "${params.requestBody.registrationChannel}",
+                   |    "accessType" : "${params.requestBody.accessType}"
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavetestadminfrontend/controllers/HelpToSaveApiController.scala]: we don't need the `accessType` in this JSON body

File: app/uk/gov/hmrc/helptosavetestadminfrontend/util/TotpGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.apache.commons.codec.binary.Base32
+
+import scala.concurrent.duration._
+import scala.math._
+
+sealed trait CryptoAlgorithm
+case object HmacSHA512 extends CryptoAlgorithm
+
+trait TotpGenerator {
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavetestadminfrontend/util/TotpGenerator.scala]: could we not bring in the TOTP generator stuff as a dependency through `build.sbt`?

File: app/uk/gov/hmrc/helptosavetestadminfrontend/util/AccessFormatter.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+object AccessFormatter {
+
+  val accessFormatter = new Formatter[AccessType] {
+    override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], AccessType] =
+      data.get(key).fold[Either[Seq[FormError], AccessType]](
+        Left(Seq(FormError(key, "There was no AccessType string found")))){
+        case "Privileged"     ⇒ Right(Privileged)
+        case "UserRestricted" ⇒ Right(UserRestricted)
+        case _                ⇒ Left(Seq(FormError(key, "Invalid AccessType found")))
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavetestadminfrontend/util/AccessFormatter.scala]: in the error it'd be helpful to know what the unrecognised AccessType string was

File: app/uk/gov/hmrc/helptosavetestadminfrontend/config/AppConfig.scala [hunk 4]
```diff
@@ -47,8 +47,11 @@ class AppConfig @Inject()(val runModeConfiguration: Configuration, environment:
     s"""{
           "client_secret":"$clientSecret",
           "client_id":"$clientId",
-          "grant_type":"authorization_code",
+          "grant_type":"client_credentials",
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavetestadminfrontend/config/AppConfig.scala]: for privileged access this is the correct value, but for user-restricted access we still need `authorization_code`

File: app/uk/gov/hmrc/helptosavetestadminfrontend/util/TotpGenerator.scala (hunk 2 above)
Comment (Jackie1988) [reply app/uk/gov/hmrc/helptosavetestadminfrontend/util/TotpGenerator.scala]: yeah I've now bought the TOTP generator stuff in via dependency and removed the trait I created

File: app/uk/gov/hmrc/helptosavetestadminfrontend/util/AccessFormatter.scala (hunk 3 above)
Comment (Jackie1988) [reply app/uk/gov/hmrc/helptosavetestadminfrontend/util/AccessFormatter.scala]: ok sure

File: app/uk/gov/hmrc/helptosavetestadminfrontend/config/AppConfig.scala (hunk 4 above)
Comment (Jackie1988) [reply app/uk/gov/hmrc/helptosavetestadminfrontend/config/AppConfig.scala]: ok

File: app/uk/gov/hmrc/helptosavetestadminfrontend/connectors/OAuthConnector.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+class OAuthConnector @Inject()(http: WSHttp, appConfig: AppConfig) extends Logging {
+
+  def requestPrivilegedAccess(totpCode: String)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Either[String, String]] = {
+    http.post(s"${appConfig.oauthURL}/oauth/token", Json.parse(appConfig.tokenRequest(totpCode))).map {
+      response =>
+        response.status match {
+          case OK =>
+            val token = (response.json \ "access_token").as[String]
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavetestadminfrontend/connectors/OAuthConnector.scala]: `as` will throw if it is not a `String` - using `validate` and handling the error case and success case separately is safer

File: build.sbt [hunk 6]
```diff
@@ -17,7 +17,8 @@ val dependencies = Seq(
   "uk.gov.hmrc" %% "play-whitelist-filter" % "2.0.0",
   "com.github.kxbmap" %% "configs" % "0.4.4",
   "org.typelevel" %% "cats-core" % "1.1.0",
-  "com.google.guava" % "guava" % "23.0"
+  "com.google.guava" % "guava" % "23.0",
+  "uk.gov.hmrc" % "totp-generator_2.11" % "0.5.0"
```
Comment (AndyHWChung) [inline build.sbt]: you don't need the `_2.11`, you can do
```
"uk.gov.hmrc" % "totp-generator" %% "0.5.0"
```
with two `%`'s instead to automatically derive the scala version

File: app/uk/gov/hmrc/helptosavetestadminfrontend/connectors/OAuthConnector.scala (hunk 5 above)
Comment (Jackie1988) [reply app/uk/gov/hmrc/helptosavetestadminfrontend/connectors/OAuthConnector.scala]: ok thanks

File: build.sbt (hunk 6 above)
Comment (Jackie1988) [reply build.sbt]: thanks

**Comment by AndyHWChung** [comment]
LGTM

**Pull request merged**